### PR TITLE
Test more of provider resolution by mocking the filesystem and commands

### LINKED ISSUE #3934: Service resource fails if systemd unit files are not present on ubuntu 15.04
```ruby
service 'go-agent' do
    action [:enable, :start]
end
```
trace
```sh
Generated at 2015-09-20 19:53:02 -0700
NoMethodError: service[go-agent] (go_cd::agent line 76) had an error: NoMethodError: undefined method `+' for false:FalseClass
/opt/chef/embedded/apps/chef/lib/chef/platform/service_helpers.rb:134:in `platform_has_systemd_unit?'
/opt/chef/embedded/apps/chef/lib/chef/platform/service_helpers.rb:98:in `config_for_service'
/opt/chef/embedded/apps/chef/lib/chef/provider/service/debian.rb:35:in `supports?'
/opt/chef/embedded/apps/chef/lib/chef/provider_resolver.rb:79:in `block in supported_handlers'
/opt/chef/embedded/apps/chef/lib/chef/provider_resolver.rb:79:in `select'
/opt/chef/embedded/apps/chef/lib/chef/provider_resolver.rb:79:in `supported_handlers'
/opt/chef/embedded/apps/chef/lib/chef/provider_resolver.rb:91:in `prioritized_handlers'
/opt/chef/embedded/apps/chef/lib/chef/provider_resolver.rb:116:in `maybe_dynamic_provider_resolution'
/opt/chef/embedded/apps/chef/lib/chef/provider_resolver.rb:61:in `resolve'
/opt/chef/embedded/apps/chef/lib/chef/resource.rb:1619:in `provider_for_action'
/opt/chef/embedded/apps/chef/lib/chef/resource.rb:582:in `run_action'
/opt/chef/embedded/apps/chef/lib/chef/runner.rb:49:in `run_action'
/opt/chef/embedded/apps/chef/lib/chef/runner.rb:81:in `block (2 levels) in converge'
/opt/chef/embedded/apps/chef/lib/chef/runner.rb:81:in `each'
/opt/chef/embedded/apps/chef/lib/chef/runner.rb:81:in `block in converge'
/opt/chef/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:83:in `block in execute_each_resource'
/opt/chef/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chef/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chef/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chef/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:81:in `execute_each_resource'
/opt/chef/embedded/apps/chef/lib/chef/runner.rb:80:in `converge'
/opt/chef/embedded/apps/chef/lib/chef/client.rb:654:in `block in converge'
/opt/chef/embedded/apps/chef/lib/chef/client.rb:649:in `catch'
/opt/chef/embedded/apps/chef/lib/chef/client.rb:649:in `converge'
/opt/chef/embedded/apps/chef/lib/chef/client.rb:688:in `converge_and_save' 
/opt/chef/embedded/apps/chef/lib/chef/client.rb:269:in `run'
/opt/chef/embedded/apps/chef/lib/chef/application.rb:270:in `block in fork_chef_client'
/opt/chef/embedded/apps/chef/lib/chef/application.rb:258:in `fork'
/opt/chef/embedded/apps/chef/lib/chef/application.rb:258:in `fork_chef_client'
/opt/chef/embedded/apps/chef/lib/chef/application.rb:224:in `block in run_chef_client'
/opt/chef/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
/opt/chef/embedded/apps/chef/lib/chef/application.rb:212:in `run_chef_client'
/opt/chef/embedded/apps/chef/lib/chef/application/client.rb:402:in `block in interval_run_chef_client'
/opt/chef/embedded/apps/chef/lib/chef/application/client.rb:392:in `loop'
/opt/chef/embedded/apps/chef/lib/chef/application/client.rb:392:in `interval_run_chef_client'
/opt/chef/embedded/apps/chef/lib/chef/application/client.rb:382:in `run_application'
/opt/chef/embedded/apps/chef/lib/chef/application.rb:60:in `run'
/opt/chef/embedded/apps/chef/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:55:in `load'
/usr/bin/chef-client:55:in `<main>'

```

### PR DESCRIPTION
**Pull request by jkeiser**
Fixes #3934, based off Ranjib's fix in #3935.

This:
1. Incorporate's Ranjib's fix to make provider resolution work in Ubuntu 12.05, where systemctl returns a nonzero exit code.
2. Modifies the ProviderResolver service tests so that they test provider resolution more deeply, mocking the filesystem and shellout commands to pretend to be various init systems. No actual changes are made to the set of tests, just how the `stub_service_*` methods are implemented.
3. [Adds a test](https://github.com/chef/chef/commit/290c9a0727b5ffa4bf47da7c00e40996c6630a3e) for when the systemctl command returns a nonzero exit code.

### CODE REVIEW

**Comment by jkeiser** [comment]
@ranjib @lamont-granquist @dan (cc @chef/client-core) could use a review on the test changes here. (See description.)

File: lib/chef/platform/service_helpers.rb [hunk 1]
```diff
@@ -126,7 +131,7 @@ def extract_systemd_services(command)
           # this splits off the suffix after the last dot to return "sshd"
           services += services.select {|s| s.match(/\.service$/) }.map { |s| s.sub(/(.*)\.service$/, '\1') }
         rescue Mixlib::ShellOut::ShellCommandFailed
```
Comment (danielsdeleo) [inline lib/chef/platform/service_helpers.rb]: Are there other exit statuses besides `1` that we should maybe not be rescuing (or treating as success, anyway)? Perhaps adding `return: [0,1]` to the `shell_out` call would be safer?

File: spec/support/shared/unit/double_world.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#
+# world.files["/etc/rc.d/init.d"] = "blah" means that /etc/rc.d/init.d is a file with contents "blah."
+# world.commands["shellme"] = "result" means that when shell_out! calls "shellme", it will return "result" in stdout.
+# world.commands["shellme"] = 10 means that when shell_out! calls "shellme", it will return exit code 10
+# world.commands["shellme"] = { code: 10, stdout: "result", stderr: "result2"}
+#
+shared_context "double world" do
+  let(:world) { DoubleWorld.new }
+  before do
+    allow(::File).to receive(:exist?)      { |f| world.files.has_key?(f) }
```
Comment (danielsdeleo) [inline spec/support/shared/unit/double_world.rb]: I'd prefer a stricter approach where you have to say which files you expect the code to try to access that _don't_ exist. Maybe you would do `world.files["/etc/rc.d/init.d"] = false` or something like that. That way you'd have to be explicit about what the state of the filesystem is and you're less likely to accidentally rely on the implicit file-doesn't-exist assumption built in to this.

Comment (jkeiser) [inline spec/support/shared/unit/double_world.rb]: @danielsdeleo I think it's doing what you expect, if I understand you right: *only* the files you specify are considered to exist. Like, `::File.exists?('/')` will actually fail on all systems with the code the way it is now (and I'm OK with that).

File: lib/chef/platform/service_helpers.rb (hunk 1 above)
Comment (jkeiser) [inline lib/chef/platform/service_helpers.rb]: @ranjib what is the exit code we want to catch anyway? Is there a correct one?

**Comment by lamont-granquist** [comment]
After the file provider unit tests, I really dislike mocking out posix filesystems.  There is simply way too many possible calls the code can make to poke the filesystem so from the standpoint of unit testing you have two systems with way too much collaboration between them.  Either you need to refactor your API to make it simpler (which you inherently can't because you have to use files), or you need to expand your definition of the system-under-test to include the filesystem.

If you look at what I did with the debian ifconfig provider, I injected real tempfiles from the specs (via define_const, which you might not like, but i thought it made clear sense to me):

https://github.com/chef/chef/blob/master/spec/unit/provider/ifconfig/debian_spec.rb#L58-L68

Then it doesn't matter what calls or call sequence the code makes, you don't have to mock anything and if the file starts with one contents and winds up with the correct contents, then you know it worked -- doesn't matter what route someone chose to get there.

For this, though, you're just checking existence and if you stub the filenames its kind of halfway pointless.  I'd suggest probably making the filenames depend on a 'chroot' variable that was '/' and then you could dep inject a tmpdir in the specs and a.  validate that the filename was the chroot + the real path, and b. validate that if it exists the method returns the right value.

IMO, we don't need to go down the road of trying to write our own MockFS/FakeFS.  That's looking at the problem wrong.

**Comment by jkeiser** [comment]
@lamont-granquist went ahead and did it as you suggested (roughly).

**Comment by jkeiser** [comment]
Lamont :+1:'d externally, closing.

**Pull request merged**